alws/build_planner.py

```diff
@@ -189,6 +190,30 @@ async def prepare_module_index(
         allowed_arches = (task_arch, 'src', 'noarch')
         index = IndexWrapper.from_template(task.modules_yaml)
 
+        # This sorting and filtering is needed due to s390x artifacts.
+        # Source RPMs there have higher module_index hence they get
+        # into modules for other architectures where they are not needed.
+        # On the other hand, sources with lesser module_index hop into s390x
+        # modules.yaml metadata which is also wrong. This filtering allows
+        # to catch those source artifacts and filter them out.
+        src_bin_mapping = defaultdict(set)
+        allowed_artifacts = set()
+        module_id_regex = re.compile('\+(?P<module_id>\d+)\+')
+        for ref in task.refs:
+            if ref.enabled or not ref.added_artifacts:
+                continue
+            for artifact in ref.added_artifacts:
+                parsed_artifact = RpmArtifact.from_str(artifact)
+                module_id = module_id_regex.search(
+                    parsed_artifact.release).groupdict().get('module_id', '')
+                if parsed_artifact.arch in allowed_arches:
+                    src_bin_mapping[module_id].add(parsed_artifact)
+
+        for id_, artifacts in src_bin_mapping.items():
+            if len(artifacts) == 1 and all([i.arch == 'src' for i in artifacts]):
```

Comment:
Can we get rid of `len(...) == 1` here?
also we can throw generator instead of list comprehension
```python3
if all((i.arch == 'src' for i in artifacts)):
    ...
```